mm2src/lp_ordermatch.rs

```diff
@@ -114,26 +115,76 @@ async fn process_order_keep_alive(
     let ordermatch_ctx = OrdermatchContext::from_ctx(&ctx).unwrap();
     let mut orderbook = ordermatch_ctx.orderbook.lock().await;
     let uuid = keep_alive.uuid.into();
-    match find_order_by_uuid_and_pubkey(&mut orderbook, &uuid, from_pubkey) {
-        Some(order) => {
-            order.timestamp = keep_alive.timestamp;
-            true
-        },
-        None => match ordermatch_ctx.inactive_orders.lock().await.remove(&uuid) {
-            Some(mut order) => {
+    if let Some(order) = find_order_by_uuid_and_pubkey(&mut orderbook, &uuid, from_pubkey) {
+        order.timestamp = keep_alive.timestamp;
+        return true;
+    }
+
+    if let Some(mut order) = ordermatch_ctx.inactive_orders.lock().await.remove(&uuid) {
+        order.timestamp = keep_alive.timestamp;
+        orderbook
+            .entry((order.base.clone(), order.rel.clone()))
+            .or_insert_with(HashMap::new)
+            .insert(uuid, order);
+        return true;
+    }
+
+    log!("Couldn't find an order " [uuid] ", try request it from peers");
+    match request_order(ctx.clone(), uuid, from_pubkey).await {
+        Ok(Some(mut order)) => {
+            if order.timestamp < keep_alive.timestamp {
+                // a peer may not received the keep_alive yet
                 order.timestamp = keep_alive.timestamp;
-                orderbook
-                    .entry((order.base.clone(), order.rel.clone()))
-                    .or_insert_with(HashMap::new)
-                    .insert(uuid, order);
-                true
-            },
-            None => {
-                broadcast_repeat_order(ctx, topic.into(), keep_alive.uuid.into());
-                false
-            },
+            }
+            insert_or_update_order(&ctx, order, uuid).await;
```

Comment:
It is a dead lock, orderbook lock is already acquired, `insert_or_update_order` also acquires it. Even if request order result is None or Err It can also cause problems because request can take significant time blocking the access to orderbook.
@sergeyboyko0791 Could you please fix this and push the fix directly to `libp2p-gossipsub-ordermatch` branch?

Reply:
Yes sure. I didn't test the lp_network and lp_oredermatch changes. I should think how to test this